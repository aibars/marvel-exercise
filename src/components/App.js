import React from 'react';
import '../styles/App.css';
import ItemsList from './ItemsList';
import SelectedItem from './SelectedItem';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  fetchItems,
} from '../actions';

class App extends React.Component {

  constructor(props) {
    super(props)
    this.handleRefreshClick = this.handleRefreshClick.bind(this);
  }

  componentDidMount() {
    const { dispatch } = this.props
    dispatch(fetchItems())
  }

  handleRefreshClick(e) {
    e.preventDefault();

    const { dispatch } = this.props;
    //dispatch(invalidateItems());
    dispatch(fetchItems());
  }

  render() {
    const { selectedItem, items, isFetching, lastUpdated } = this.props;
    return (
      <div className="App" style={{ opacity: isFetching ? 0.5 : 1 }}>
        <header className="App-header">
          Master-Detail Exercise
        </header>
        <div id="left-col">
          <label>Pokemons List:</label> {!isFetching && (
            <button className="refresh-btn" onClick={this.handleRefreshClick}>Refresh</button>
          )}
          <br />
          <span className="update-info">
            {lastUpdated && (
              <span>
                Last updated at {new Date(lastUpdated).toLocaleTimeString()}.{' '}
              </span>
            )}
          </span>
          {isFetching && items.length === 0 && <h2>Loading...</h2>}
          {!isFetching && items.length === 0 && <h2>Empty.</h2>}
          {items.length > 0 &&
            <div >
              <ItemsList items={items} />
            </div>
          }
        </div>
        {selectedItem && <SelectedItem />}
      </div >
    );
  }
}

App.propTypes = {
  selectedItem: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
  isFetching: PropTypes.bool.isRequired,
  lastUpdated: PropTypes.number,
  dispatch: PropTypes.func.isRequired
}

function mapStateToProps(state) {
  const { selectedItem, pokemons } = state;
  const { isFetching, lastUpdated, items } = pokemons;

  return {
    selectedItem,
    items,
    isFetching,
    lastUpdated
  };
}

export default connect(mapStateToProps)(App);