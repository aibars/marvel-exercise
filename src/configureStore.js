import { createStore, applyMiddleware, compose } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import rootReducer from './reducers';

const loggerMiddleware = createLogger();

export default function configureStore(preloadedState) {
  const enhancers = [];
  const windowIfDefined = typeof window === 'undefined' ? null : window;
  if (windowIfDefined && windowIfDefined.__REDUX_DEVTOOLS_EXTENSION__) {
    enhancers.push(windowIfDefined.__REDUX_DEVTOOLS_EXTENSION__());
  }


  return createStore(
    rootReducer,
    preloadedState,
    compose(applyMiddleware(thunkMiddleware, loggerMiddleware), ...enhancers),
  );
}