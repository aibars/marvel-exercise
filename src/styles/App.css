.App-header {
  background-color: #282c34;
  min-height: 100vh;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

#left-col {
  height: 100%;
  float: left;
  width: 50%;
}

#right-col {
  width: 50%;
  height: 100%;
  float: right;
}