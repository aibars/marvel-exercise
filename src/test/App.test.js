import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import App from '../components/App';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);

describe('App', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      pokemons: {
        items: [],
        isFetching: false,
        lastUpdate: Date.now()
      },
      selectedItem: {
        item: null,
        isFetching: false
      },
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );

  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch an action on refresh click', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(3);
  });
});