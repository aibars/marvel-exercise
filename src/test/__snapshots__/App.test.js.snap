// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render with given state from Redux store 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    Master-Detail Exercise
  </header>
  <div
    id="left-col"
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <label
      className="list-title"
    >
      Pokemons List 
    </label>
    <button
      className="refresh-btn"
      onClick={[Function]}
    >
      Refresh
    </button>
    <h2>
      Empty.
    </h2>
  </div>
  <div
    id="right-col"
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <label
      className="list-title"
    >
       Selected Pokemon:
    </label>
  </div>
</div>
`;
